Improved platform tests (#1291)
openshift/enhancements: enhancements/testing/improved-platform-tests.md

- OpenShift 4.10 / kubernetes 1.23 - 7 days (compare [o/k#1087](https://github.com/openshift/kubernetes/pull/1087)
  vs [origin#26753](https://github.com/openshift/origin/pull/26753)).
- OpenShift 4.11 / kubernetes 1.24 - 21 days (compare [o/k#1252](https://github.com/openshift/kubernetes/pull/1252)
  vs [origin#27244](https://github.com/openshift/origin/pull/27244)).
- OpenShift 4.12 / kubernetes 1.25 - 40 days (compare [o/k#1360](https://github.com/openshift/kubernetes/pull/1360)
  vs [origin#27526](https://github.com/openshift/origin/pull/27526)).

Moreover, in the past two releases the changes to origin repository were reverted
at least once, due to multiple issues found, which also affected the increased
delays.

### User Stories

- As an OpenShift developer, I want to be able to seamlessly run OpenShift
  and kubernetes-specific tests using `openshift-tests run`.
- As an OpenShift developer, I want to be able to run newest kubernetes tests
  without the necessity to bump vendored code.
- As an OpenShift developer, I want to be able to extend and monitor cluster
  during tests execution.
- As an OpenShift user, I want to be able to verify that my cluster is providing

>   the desired functionality running one of the provided test suites.

Reviewer: This sounds really interesting. 

A complication that would be nice to solve, we have to be careful with test names, they are the sole identifier in all our tooling. Changing test names effectively makes a new test and disconnects from all past history, skips in the release repo, jira's linked to the old test name, etc. This includes adding new [tags] (I think this should get fixed though)

We're not doing great on this even in origin today, but in separate repos we won't have a chance unless we can figure something out for warning devs they're about to cause some mayhem if a test is renamed.

Still a worthwhile goal though.

>   the desired functionality running one of the provided test suites.

Reviewer: It is quite common yes, biggest example lately was the [apigroup] tags coming in for flex installs, and with the way ginko lets you nest test names a change at one level can impact all the tests under it, things like that make it pretty common and our number of tests in the db has grown from about 5k when we started using a db in sippy this summer, to 40k now, mostly due to renames.

I'd actually like us to start extracting tags from test names but that poses a lot of problems for test linking to jiras, skips in release repo, etc.

### Goals

1. Reduce the risk updating kubernetes caused by excessive co-located code.
2. Minimize time when we start running newest kubernetes tests.
3. Maintain `openshift-tests` API for running tests.

### Non-Goals

1. Abandon existing tests.

> 2. Change the available functionality of `openshift-tests`.

Reviewer 2: IIUC it is not a goal to change the current content of `openshift-tests` suites, mainly the conformance ones (`openshift/conformance` and `kubernetes/conformance`), is that correct?

My concern is the e2e tests from kubernetes/K8s suite are in `openshift/conformance`, this suite is a base suite used to validate external clusters by partners on VCSP program - and we understand that the e2e tests in this suite cover all needed to validate an OCP installation.
```
$ ./openshift-tests run openshift/conformance --dry-run |grep -F "$(./openshift-tests run kubernetes/conformance --dry-run |tail -n1)"
"[sig-storage] Subpath Atomic writer volumes should support subpaths with secret pod [Excluded:WindowsDocker] [Conformance] [Suite:openshift/conformance/parallel/minimal] [Suite:k8s]"
$ ./openshift-tests run openshift/conformance --dry-run |grep -F "$(./openshift-tests run kubernetes/conformance --dry-run |head -n1)"
"[sig-api-machinery] AdmissionWebhook [Privileged:ClusterAdmin] listing mutating webhooks should work [Conformance] [Suite:openshift/conformance/parallel/minimal] [Suite:k8s]"
$ ./openshift-tests run openshift/conformance --dry-run |grep -F "$(./openshift-tests run kubernetes/conformance --dry-run |shuf |head -n1)"
"[sig-api-machinery] Watchers should observe an object deletion if it stops meeting the requirements of the selector [Conformance] [Suite:openshift/conformance/parallel/minimal] [Suite:k8s]"

```

Reviewer 3: everything you said is true, i'm not clear what the concern is though?

we can meet the goal of having k8s tests run as part of the openshift/conformance suite w/o including the k8s tests in the openshift-tests binary.

Reviewer 2: > we can meet the goal of having k8s tests run as part of the openshift/conformance suite w/o including the k8s tests in the openshift-tests binary.

That's what I would like to make sure I got correctly to track changes on the VCSP program/OPCT.

Overall the two steps in this proposal will bring a huge benefit.

Author: @mtulio correct, we will ensure the current workloads continue working, at least wrt k8s conformance tests.

> 3. Minimize memory footprint for running `openshift-tests`.

Reviewer 3: 4. moving the upgrade+monitoring tests that live in openshift-tests itself, out of that binary.

Reviewer: Can you elaborate on this one @bparees.

Reviewer 3: The code for the openshift-tests binary defines some "virtual" tests (they aren't ginkgo tests, they are tests that are run/managed in code and directly spit out junit results, iirc).  Those tests need to be part of the wrapper openshift-tests binary (vs just being another e2e test/binary that gets invoked/wrapped) because they either:

1) need to run during the entire duration of the test job (monitoring the cluster as other tests are run)
2) orchestrate upgrades and specific tests/verifications performed at various points before/during/after the upgrade

So given the way those tests need to be orchestrated/executed/managed they don't fit as a traditional e2e test that is just part of a ginkgo suite.  

That is why I wanted to call out that it is a non-goal to move those out of the openshift-tests wrapper.  They will need to stay where they are (or if they're going to move, that's going to be a different conversation about how to move them and where to move them to, and probably not something we want to hold this EP up over)

Reviewer 3: marking this resolved since i see it was added as item 4.

Author: It's a non-goal, but I've added this and the split of k8s tests as "Even more future work"

## Proposal

Currently, `openshift-tests` when running any test suite using `run` command,
does a few things:
1. Discovers the deployment topology and sets up all the arguments necessary
   to monitor the cluster and run the tests.
2. Starts monitoring the cluster functionality.
3. Iterates over all the tests, creating a subprocess invoking `openshift-tests run-test`
   passing a single test to run as the argument, and reports back the result to the main
   process.

This proposal's goal is to replace the internal logic (step 3 from above) with

> direct call to [ginkgo](https://onsi.github.io/ginkgo/), similarly how kubernetes

Author: Nope, nothing changes wrt forking.

runs its [e2e tests](https://github.com/kubernetes/kubernetes/blob/release-1.26/hack/ginkgo-e2e.sh).
To achieve this, the work is divided into two steps:

1. The first step aims at replacing `openshift-tests run-test` invocation with a direct
   ginkgo call.

> 2. The followup work would look at running tests in groups (as currently defined in origin),

Reviewer: +1 on both of these from my PoV. I and several others have found the current system difficult to grok and abnormal, a more standard method of invoking tests would be nice. 

When I've raised this however concerns were raised that our tests do not properly clean up after themselves, and moving them all in process would potentially introduce cross contamination especially now that they've been done in separate processes for so long. I would think most of the concern here is things left in the cluster, which running in separate processes doesn't help. But I am not fully clear as to what was seen to lead to the introduction of run-test. Any thoughts around how we will know or ensure this is safe bringing things back to running under ginko?

Reviewer 4: About benefits of doing #2, I have the impression it could save lots of resources. We need to measure to be sure, but I've seen jobs getting OOM-killed pretty often (we had to increase the job limits to "fix" that). I think it's worth considering it.

Author: That part will actually not change much, due to cleaning and trying to maintain the change minimal :/

   rather than invoking them one-by-one.

### Implementation Details/Notes/Constraints

To achieve the first step of the proposal we need to:
1. Build ginkgo and test binary (`origin-e2e.test`) containing all OpenShift tests

>    (see [proof of concept](https://github.com/openshift/origin/pull/27570))).

Reviewer 2: it would be nice to note the results (total time, suite execution summary [pass,failures...]) of this PoC [when calling `ginko` instead `run-test`](https://github.com/openshift/origin/pull/27570/files#diff-c4d33d1773570415bcef7a544d0ade716203cfb555f4ed5085d6cc768146ac56L291-R310)

Author: I've added a note in the risks and mitigations section about the additional time needed to extract binaries (based on the updated design) and `OPENSHIFT_SKIP_EXTERNAL_TESTS`, which provided a fall-back.

> 2. Ensure that `openshift-tests` deployment discovery is properly passed to `ginkgo`

Reviewer 3: https://github.com/openshift/origin/pull/27563 proposes a model where there's a single origin openshift-tests wrapper binary that has the existing CLI, but which:

1) discovers various test binaries provided by components/repos
2) interrogates them for the list of tests they provide
3) constructs the finalized list of tests to invoke and which binary each test is associated with
4) runs the tests (invoking the appropriate binary for each test)

Which i think is more or less what you want....but is also more or less what i think @soltysh is proposing here.  

The primary difference between the two is that my prototype assumes that each repo has its own lightweight wrapper binary (not the full openshift-tests, just a limited binary that reports on what tests it provides, and implements the `run-test` piece of openshift-tests).  Whereas I think Maciej's proposal would do away w/ the per repo lightweight wrapper binary and instead try to just rely the origin wrapper invoking ginkgo-built binaries directly.  So you'd still have the existing openshift-tests cli experience with the possible exception of `run-test` (the origin wrapper could still implement `run-test` though, i believe, if we wanted it)

> 2. Ensure that `openshift-tests` deployment discovery is properly passed to `ginkgo`

Author: After the initial work on the POC, it turned out that directly invoking ginkgo isn't feasible due to:
- us requiring to tweak the e2e namespace, since all of k8s tests assume they are running with full privileges, see [createTestingNS](https://github.com/openshift/origin/blob/bfcc30fde82a91fab1c881a836e4ad8e63c72fb5/test/extended/util/test.go#L169) which is responsible for that. 
- lack of ability to list tests in parsable form, `ginkgo --dry-run` does not have that option, and relying on parsing non-structured test if always problematic in maintanace
- the requirement for custom annotations, which allow us to control when and which tests are being run.

Based on the above I had to fallback to the approach from Ben, and created a minimal wrapper which handles the above points.

> 2. Ensure that `openshift-tests` deployment discovery is properly passed to `ginkgo`

Reviewer 3: another reason i think avoiding the wrapper would be difficult/impossible is the need for the "long running" tests the monitor the cluster either over the whole run or during specific time periods like upgrades.

not sure how we'd have achieved that with just ginkgo primitives and no coordinating wrapper.

   invocation.
3. Ensure that `openshift-tests` understands the output from running tests using
   `ginkgo` (either by ensuring similar output, or by updating the wrapper accordingly).
4. Build the tests image such that it will contain both `origin-e2e.test` and
   `k8s-e2e.test` coming from [our o/k fork](https://github.com/openshift/kubernetes/blob/master/openshift-hack/test-kubernetes-e2e.sh).

To achieve the second step of the proposal we need to:
1. Allow ginkgo to specify a list of tests, rather than a single regular expression.
2. Change the `openshift-tests` internal executor to run a group of test using
   `ginkgo`.

### Risks and Mitigations

> TBD

Reviewer 4: The current policy when the o/o rebase PR breaks the CI is to **revert the o/o PR**.

This policy will be harder to maintain with this change because reverting the o/k rebase PR can be problematic.

There should be fewer reasons to do any reverting because we'll notice the CI broken _during_ the o/k rebase PR, but that might not always be the case.

CC @dgoodwin

Reviewer 5: Some of the work that has been done to build the partner certification tool will need to be updated to pull additional images with separate test binaries. I think sonobuoy already supports something like that, but we should get @mtulio's input.

### Drawbacks

The major issue with splitting the test binaries to be built from two separate
repositories is that developers wishing to run kubernetes tests will have to
build the binary manually, from a separate repository and ensure the binary

> is available during test runs.

Reviewer 3: or just pull down the latest tests image.

Reviewer 2: > developers wishing to run kubernetes tests will have to
build the binary manually, from a separate repository

@bparees it's not clear to me how it affects the suite `kubernetes/conformance` currently included in the openshift-tests binary when running:
~~~
openshift-tests run kubernetes/conformance [args...]
~~~

Will the suite `kubernetes/conformance` keeps available without additional steps or we'll need to run it from the upstream way (building/downloading the binary [, using sonobuoy's embedded k8s suites], etc) ?

Reviewer 3: really a question for @soltysh at this point.  The current implementation includes the k8s tests in the openshift-tests image *and* in a separate k8s image, so you can run them from either location (i.e. just having openshift-tests is sufficient to run the k8s tests, though the version of the k8s tests may vary between the two images).  

But that is currently done by vendoring the k8s tests into origin, which ultimately we want to stop doing.  So i think the end solution looks more like:

separate test images for different tests/suites/components/areas
an aggregated image (openshift-tests) that is built by pulling all those separate images together

and only the aggregated image ships in the payload, and it's all that is needed to run all the tests.  That would maintain the current end user experience.

but that is not what has been built/delivered so far, out of a need for expediency.

Author: @mtulio there are 2 possible ways with the current approach:
1. run `openshift-tests` as you did before, which will try to pull the appropriate `k8s-tests` binary from the release image, and it will use that binary for the test execution;
2. use `OPENSHIFT_SKIP_EXTERNAL_TESTS=true` env variable, when invoking `openshift-tests` which will use the embedded  version of k8s conformance tests; 

We are guaranteeing that the latter approach will also work for cases when pulling the binary from release is not possible for various reasons. When it comes to k8s conformance (which don't change between minor versions) that shouldn't be a problem for you. 

Have a look at `Risks and Mitigations` section, if you don't find answers there, lemme know I'll gladly add whatever details are required.

   (see [proof of concept](https://github.com/openshift/origin/pull/27570))).
2. Ensure that `openshift-tests` deployment discovery is properly passed to `ginkgo`
   invocation.
3. Ensure that `openshift-tests` understands the output from running tests using
   `ginkgo` (either by ensuring similar output, or by updating the wrapper accordingly).
4. Build the tests image such that it will contain both `origin-e2e.test` and
   `k8s-e2e.test` coming from [our o/k fork](https://github.com/openshift/kubernetes/blob/master/openshift-hack/test-kubernetes-e2e.sh).

To achieve the second step of the proposal we need to:
1. Allow ginkgo to specify a list of tests, rather than a single regular expression.
2. Change the `openshift-tests` internal executor to run a group of test using
   `ginkgo`.

### Risks and Mitigations

> The current policy when the o/o rebase PR breaks the CI is to revert the o/o PR.
> This policy will be harder to maintain with this change because reverting the o/k rebase PR can be problematic.
> There should be fewer reasons to do any reverting because we'll notice the CI broken during the o/k rebase PR, but that might not always be the case.

> > Some of the work that has been done to build the partner certification tool will need to be updated to pull additional images with separate test binaries. I think sonobuoy already supports something like that, but we should get @mtulio's input.

Reviewer 2: The provider validation tool (OPCT) uses Sonobuoy to orchestrate the test environment, implementing a custom [Sonobuoy plugin](https://github.com/vmware-tanzu/sonobuoy-plugins) to run OpenShift conformance suites, and other tests/validations. The plugin uses `openshift-tests` in the following ways:
- 1: extracts, in runtime, the `openshift-tests` binary from the `tests` image from the internal registry, so we don't need to create any dependency/versions/vendor for the origin repo, and can run only the tests on o/k suites added for the release in the installed cluster when running the tool
- 2: runs `run kubernetes/conformance [...]` suite and collects the junits passing it to Sonobuoy's processor flow
- 3: runs `run openshift/conformance [...]` suite and collects the junits passing it to Sonobuoy's processor flow
- 4: runs `run-upgrade none [...]`, when validating the upgrade path, and collects the junits passing it to Sonobuoy's processor flow

The 2 can be replaced by Sonobuoy's [built-in e2e (k8s)](https://sonobuoy.io/docs/v0.56.16/e2eplugin/) plugin if the goal of this EP is to remove the kube/conformance suite/tests from the openshift-tests binary. But having the possibility to access the tests without depending on sonobuoy would be something nice to have. Do you know if the kube test binary is published somewhere/image within the cluster?

I created the card https://issues.redhat.com/browse/OPCT-211 to investigate it more deeply.

Reviewer 3: > extracts, in runtime, the openshift-tests binary from the tests image from the internal registry

this is something that may not work when we get to the goal state, which is that we'll have multiple binaries (and not just "openshift-tests" and "k8s-tests" but a binary per component).  Will it be a problem if the plugin needs to extract an arbitrary set of binaries from the tests image?  the openshift-tests binary would be one of them and would be responsible for invoking the other test binaries, so the sonobouy plugin would just need to extract all the binaries and place them in a spot where the openshift-tests binary would find them when invoked.

Reviewer 2: > Will it be a problem if the plugin needs to extract an arbitrary set of binaries from the tests image?

Not a problem, but would be nice to have the list of the binaries that need to be extracted dynamically from a given suite, so we can iterate over it and extract it from the `tests` image. Maybe a command of openshift-tests? Something like:

~~~
$ openshift-tests run suite-name --dry-run --dependencies
cluster-machine-api-operator.test
cluster-image-registry-operator.test
~~~

does it make sense?

Reviewer 3: i don't think we want openshift-tests to have to include a hardcoded list of binaries, the idea is for it to be able to dynamically invoke all the binaries that are provided.  but we can definitely work something out in terms of providing the list of binaries, if nothing else as a text file in the image itself.

Reviewer 2: > if nothing else as a text file in the image itself
it works too, we just need to know which binary is required to run the conformance suites correctly/successfully.

Another alternative is to teach the plugin to use the image `tests` itself, instead of extracting the binary, assuming all binaries are already inside the `tests` image and resolved when running the o/k conformance suites.

Author: `openshift-tests` will definitely not have that information available, we're planning to encode that information in the release image, see "Future work" section of the proposal. Adding a sub-command which will give you a list of external binaries might be possible, since we're planning to annotate images in the release payload with information which ones have additional test binaries, and where in the image they are located. 

Having said that, I'd probably leave the extraction within `openshit-tests` since it will know which binaries and where to retrieve them from. They will definitely be in several images, so you'd need to know both how to extract that information from release image and where to look for them, which `openshift-tests` will know how to do.

All in all, based on what you wrote above, I'm convinced that even in the case of running `openshift-tests` in a valid openshift cluster it can easily figure out what it needs and where to get it from, assuming the release image is reachable from where you're running the `openshift-tests` binary.

Reviewer 3: > They will definitely be in several images

why?  why did we give up on the possibility of producing a single test image that contains all the test binaries?  From a consumption perspective that seems preferable and more reliable.

Author: CI limitation, the current workflow doesn't make that easy afaict. I've added that option as one of the considered alternatives.

---
title: improved-platform-tests
authors:
  - "@soltysh"
reviewers:
  - "@ardaguclu"
  - "@bertinatto"
  - "@deads2k"
  - "@ingvagabund"
approvers:
  - "@deads2k"
api-approvers:
  - None
creation-date: 2022-11-16
last-updated: 2023-05-31
tracking-link:
  - https://issues.redhat.com/browse/WRKLDS-720
replaces:
  - "enhancements/testing/extended-platform-tests.md"
---

# Improved Platform Tests

## Summary

Allow maintaining and building end-to-end (e2e) tests outside of
[origin](https://github.com/openshift/origin/) repository, and running them under
single `openshift-tests` umbrella, by extracting the necessary test binaries
during runtime.

## Motivation

> The reasoning behind this change are two-folded:

Reviewer 3: ```suggestion
The reasoning behind this change are two-fold:
```

> 1. Short term goal is to speed up the [kubernetes updates](#kubernetes-updates).

Reviewer 3: ```suggestion
1. Short term goal is to speed up the [kubernetes updates](#kubernetes-updates) by enabling us to test the proposed rebase against the updated upstream tests instead of old versions of them"
```

2. Long term goal is separating all e2e tests so that they can live in their
   [respective component repos](#tests-co-located-with-code).

### User Stories

- As an OpenShift developer, I want to be able to seamlessly run OpenShift
  and kubernetes-specific tests using `openshift-tests run`, so that my current
  workflows continue working as before.
- As an OpenShift developer, I want to be able to run newest kubernetes tests
  without the necessity to bump vendored code, to limit the amount of re-vendoring

>   work and ensure all proper test coverage.

Reviewer 3: ```suggestion
  work and ensure all proper test coverage/alignment of the version of the code being tested with the version of the tests.
```

- As an OpenShift developer, I want to be able to extend and monitor cluster
  during tests execution, to ensure high availability of the platform.
- As an OpenShift partner or user, I want to be able to verify that my cluster is providing
  the desired functionality (CSI for example) running one of the provided test suites.
- As an OpenShift developer, I want to be able to deliver new e2e tests in the
  same pull request (PR) that introduces the feature being tested, so that I can
  ensure my new feature code is properly tested prior to merging.

### Goals

1. Reduce the risk updating kubernetes caused by excessive co-located code,
   and minimize time when we start running newest kubernetes tests.
2. Remove the requirement that all test code lives in a single repository.
3. Maintain `openshift-tests` API for running tests.
4. Ensure that kubernetes tests are still vendored in `openshift-tests` binary,

>    to allow for a fallback when the release payload image is not reachable.

Reviewer 3: I see this one as temporary....once we start having all teams moving their e2e tests into their own repos, we're just going to have to accept that the test images need to be available or the test job fails.

and we're not going to want to maintain this vendoring indefinitely just to give us a fallback (not to mention that falling back is risky since now you're not necessarily running the tests you thought you were)

Author: For some cases the fallback will be a must, microshift comes to my mind immediately, since they don't provide release image. I'm not sure about hypershift, either. So at least for k8s-related tests I think that might be longer than just temporary. We'll need to re-assess this at a later point in time. That's why I've explicitly called out k8s tests only for backwards compatibility.

Author: I've added this to "Even more future work" section, there's still work about upgrade tests, so this is a good future thing to look at .

Reviewer 3: if nothing else, fallback needs to be an opt-in behavior (do not fallback to another version of the test code silently or by default).

but this is another reason why producing a single image that contains all the test binaries would be a better solution in my mind.

cc @dhellmann since he can better speak to what microshift does today or would want to support in the future.

Author: Yeah, that explicit fallback is being tracked in https://github.com/openshift/origin/pull/28000 which will eventually allow us to drop the fallback we currently have.

sequenceDiagram
    actor U as User
    participant O as openshift-tests
    participant E as external-tests
    participant R as release:tag
    U->>+O: run openshift/conformance
    O->>+O: list tests
    O->>+R: extract test binary
    rect rgb(177, 177, 177)
        opt test binary exists?
            R->>+E: external-tests
            O->>+E: list tests
            E->>+O: tests
        end
    end
    loop over all tests
        alt embedded test?
            O->>+O: run #lt;test name#gt;#10;
        else external test?
            rect rgb(177, 177, 177)
                O->>+E: run #lt;test name#gt;
                E->>+O: results
            end
        end
    end
    O->>+U: results
-->

### API Extensions

N/A

### Implementation Details/Notes/Constraints

To achieve the first step of the proposal, we need to:
1. Provide a library for building `<binary>-tests` which exposes two commands:
   - `list` - responsible for listing tests in JSON format;
   - `run` - run a single test, returning results in ginkgo compatible format;
2. Provide an image annotation or similar mechanism informing where the test

>    binary is embedded in the release payload.

Reviewer 3: the library also needs to help teams:
1) indicate+maintain labels/tags that should be dynamically added to their tests (rules.go/annotations.go)
2) manage the use of disconnected images (i.e. ensure they are only referencing images that are mirrored, at least for conformance tests)

Author: Done.

Reviewer 3: not seeing where this got added (maybe you haven't pushed yet)

   should have an associated binary name, which will be used during invocation.
3. Expand the current filtering, shuffling, and randomization logic to also take
   into account the new tests coming from external binaries.
4. Add logic responsible for iterating over all tests, invoking the `run`
   subcommand using the associated test's binary and reading its results.
5. Provide a fallback path allowing to use the vendored tests, only when either
   the release image is not available, or on demand (`OPENSHIFT_SKIP_EXTERNAL_TESTS`
   environment variable, for example).

### Risks and Mitigations

- The current policy when the [origin](https://github.com/openshift/origin/) pull
  request bringing the updated k8s version breaks the CI is to revert PR. This
  policy will be harder to execute with this change because reverting the o/k
  rebase PR can be problematic and was never tried before.

There should be fewer reasons to do any reverting, because with this new mechanism
we should get the appropriate signal during extensive payload testing required
to land every k8s bump PR.

- Some of the work that has been done to build the partner certification tool will

>   need to be updated to pull additional images with separate test binaries.

Reviewer 3: this seems to point to a design aspect that isn't clearly specified yet.

Is the intent to have each component/externalized test binary go into its own image?

Or is the intent to bring all the externalized test binaries into the existing `tests` image? 

I prefer the latter, though it's more effort at build time...but it's less effort for teams to adopt (they don't need to start publishing a new image), doesn't grow the payload size as much, and avoids the problem of not being able pull certain images.

it also makes the list of test binaries obvious:  the `tests` image could contain a list of binaries, or even just a well known directory containing all the binaries.  If we maintain them in separate images, then something needs to make it clear what the list of component-test images is.

Author: The goal of this specific point is to ensure folks building on top of `openshift-tests` that all the additional work will be handled within binary and no additional work should be required by consumers.

Reviewer 3: > The goal of this specific point is to ensure folks building on top of openshift-tests that all the additional work will be handled within binary and no additional work should be required by consumers.

in that case i think you should just state that assurance generically.  As it's currently written it assumes a specific implementation (image per test binary).

Author: Updated.

The `openshift-tests` binary will contain the necessary logic how to retrieve those
additional tests binaries from the release payload, so even in the offline installation
as long as the release payload is reachable, the tests should continue working.

- Not available release image can prevent from running the tests.

The proposed mechanism will contain a fallback path, which will ensure that in
cases where the release payload is not reachable we will be able to run all the
tests bundled within the `openshift-tests` binary. This includes the basic

> OpenShift and Kubernetes conformance tests.

Reviewer 3: long term i don't see us keeping the k8s tests in origin, that would mean extra rebase work.

also if there is a fallback path, it needs to be opt-in and very clear that it's been taken, because it means a significant change in test coverage when falling back.  (To the point that i'm not even sure we'd ever want automatic fallback, even with a "--allow-fallback" argument.  I'd rather setup two jobs, one which uses external content and one which doesn't, rather than have a single job that sometimes falls back and sometimes doesn't)

Author: I'm willing to remove the automatic fallback, since like you're describing this will force you to be explicit about running vendored k8s vs external.
As for your worry about diverging tests, I'm less worried about that, major bulk of test changes is coming from the major k8s bump, even though patch releases have some fixes, they are rather cosmetic from what I've seen over the years.

Author: https://github.com/openshift/origin/pull/28000 removes the automatic fallback. I'll update the proposal accordingly.

which bypasses the mechanism and always uses the tests bundled in `openshift-tests`
binary.
At a later point in time, it's reasonable to re-use test binaries downloaded
in prior executions, through exposing an environment variable,
`OPENSHIFT_EXTERNAL_TESTS_DIRECTORY` for example.

- In some cases, after landing newer version of kubernetes it is possible that
  we'll be having issues with tests relying on newer kubelets, which is usually
  shipped with updated base image with some delay.

This problem will not present during PR testing, because then we always build
the base image from scratch to ensure all components are using newest versions
for tests. For those rare cases where tests will rely on newer functionality in
kubelet we can either temporarily disable the test before landing the PR, or
sync with ART to ensure the updated kubelet is shipped as quickly as possible.

- Upgrade tests, unlike other tests are executed within single test binary
  making it harder to migrate to use external test binaries.

The current upgrade test suite runs only a handful of tests which don't change
that often. For that reason, we've initially decided to rely only on the vendored

> code for the upgrade tests.

Reviewer 3: is this upstream/k8s vendored code?

Author: Yes, the way it's being run is we create a single test which wraps all the upgrade tests. Each upgrade test has to match [Test](https://github.com/kubernetes/kubernetes/blob/d9714078f821ccae1e2d50ea69c081d73a255f1f/test/e2e/upgrades/upgrade.go#L48) interface which allow you to setup the environment for the test, execute the test, and provide a teardown logic. So extracting this logic to an external binary will require a bit more effort.

### Drawbacks

This proposal might lead to unnecessary proliferation of external tests binaries,
which might cause problems when the release image is not available, or for local

> development.

Reviewer 3: hard to imagine a situation where the release image is not available to someone who's trying to run tests

Author: What about microshift?

Reviewer 3: i'll have to defer to @dhellmann on how they manage this today (where they get the openshift-tests binary from, how they distribute it to customers) on the implications there, but in my mind it's another argument for why we're better off having a single test image that contains all the test binaries.  Microshift can then consume/reference that image.

Reviewer 5: There is no release payload for MicroShift. It looks like the CI job that uses openshift-tests pulls it from an image built as part of the CI system? https://github.com/openshift/release/blob/master/ci-operator/config/openshift/microshift/openshift-microshift-main.yaml#L47

@pacevedom, @pmtk , or @copejon  can any of you explain where that binary comes from?

We do not ship the openshift-tests binary as part of MicroShift.

Reviewer 6: Following the breadcrumbs it's: inputs.test-bin which is 
```
  test-bin:
    name: "4.14"
    namespace: ocp
    tag: tests
```
Which is imported: `Tagging ocp/4.14:tests into pipeline:test-bin`.
I think it's [this image config](https://github.com/openshift/release/blob/master/ci-operator/config/openshift/origin/openshift-origin-master.yaml#L26-L32)

tl;dr: we just take it from a promoted image

Reviewer 5: So we may be OK in CI. How would someone run the tests by hand, like if they were debugging a problem?

Reviewer 6: Not sure, I'd build `openshift-tests` from source. Maybe @pacevedom has better idea

Reviewer 5: Based on this proposal from @soltysh , building from source wouldn't give us the tests that had been moved out of the repo.

@soltysh or @bparees, do you have something in mind for the workflow where someone is trying to test outside of our CI system and has no release payload?

Reviewer 3: >  the workflow where someone is trying to test outside of our CI system and has no release payload?

if they aren't having to modify the tests themselves, they could still use the test binaries from a recent payload image(since @soltysh's current proposal is that there would be multiple test images, one per binary, i defer to him on exactly how a consumer/client identifies all the images needed, pulls them, and extracts all the binaries, but that's something that CI will need to be doing too and i expect it will be relatively straightforward and handled by the openshift-tests binary itself)

if they are modifying the tests locally, then i'd expect them to be able to:
1) build/get a copy of the openshift-tests wrapper binary
2) invoke it while pointing it at their local externalized test binary for their component

(it would only run the tests that are built into openshift-tests plus the tests in their binary, but usually when people are doing this they are already narrowing the set of tests down to a few specific ones anyway).

Do those two options cover your use cases?

Author: Number 2 is covered in risks sections, where I'm talking about possibility to use locally available binaries instead of pulling them from release. It's not implemented, but it's rather simple. I'll add it to future work section.

Author: The only downside is you need to know which binaries you care about, but in case of microshift it's rather simple b/c you'll most likely care about k8s mostly.

Author: Based on [Fabio's comment](https://github.com/openshift/enhancements/pull/1291#discussion_r1246863838) we'll stick with in-binary k8s tests for now, and we'll slowly work our way out in the future, remembering about cases like CSI and Microshift.

- OpenShift 4.11 / kubernetes 1.24 - 21 days (compare [o/k#1252](https://github.com/openshift/kubernetes/pull/1252)
  vs [origin#27244](https://github.com/openshift/origin/pull/27244)).
- OpenShift 4.12 / kubernetes 1.25 - 40 days (compare [o/k#1360](https://github.com/openshift/kubernetes/pull/1360)
  vs [origin#27526](https://github.com/openshift/origin/pull/27526)).
- OpenShift 4.13 / kubernetes 1.26 - 20 days (compare [o/k#1432](https://github.com/openshift/kubernetes/pull/1432)
  vs [origin#27738](https://github.com/openshift/origin/pull/27738)).
- OpenShift 4.14 / kubernetes 1.27 - 15 days (compare [o/k#1558](https://github.com/openshift/kubernetes/pull/1558)
  vs [origin#27894](https://github.com/openshift/origin/pull/27894)).

Moreover, the changes to origin repository were reverted at least once, due to multiple
issues found, which also affected the increased delays.

By separating OpenShift tests from the ones coming from kubernetes we will:
- shorten the time when we start running most up-to-date tests from kubernetes;
- run all tests coming from kubernetes fork, especially those introduced in patch releases;

> - start getting the correct single already during [kubernetes update process](https://github.com/openshift/kubernetes/blob/master/REBASE.openshift.md).

Author: results - fixed

2. Filters tests based on: desired suite, configuration/environment of the cluster
   being tested, parallel/serial requirements.
3. Discovers the deployment topology and sets up all the arguments necessary
   to monitor the cluster and run the tests.
4. Starts monitoring the cluster functionality.
5. Iterates over all the tests, creating a subprocess invoking `openshift-tests run-test`
   passing a single test name to run as the argument, and reports back the result
   to the main process.
6. Present the summary results of running tests.

Below diagram shows the current flow:

![Old Flow](./old-flow.png)
<!--
Above diagram was generated at https://mermaid.live/ using following code:
sequenceDiagram
    actor U as User
    participant O as openshift-tests
    U->>+O: run openshift/conformance
    O->>+O: list tests
    loop over all tests
        O->>+O: run #lt;test name#gt;
        O->>+U: results
    end
-->

### Workflow Description - new

This proposal's goal is to extend test discovery (step 1), and running the actual
test (step 5) with external binaries, which are retrieved from the payload image.
The improved flow would be as follows:

> 1. Retrieve test binaries from the payload image.

Reviewer 3: this really means retrieve the set of test image(s) referenced by the payload image, and then retrieve the binaries from those images, right?  Since you seem to be assuming each binary goes in its own image (i'm still not sold on that, for the record :) )

Author: Good point - updated.

> (i'm still not sold on that, for the record :) )

it's an ongoing effort, I'll figure out a way to convince you :wink:

N/A

#### Tech Preview -> GA

N/A

#### Removing a deprecated feature

N/A

### Upgrade / Downgrade Strategy

N/A

### Version Skew Strategy

N/A

### Operational Aspects of API Extensions

N/A

#### Failure Modes

N/A

#### Support Procedures

N/A

## Implementation History

- *2022-11-16* - Initial draft.
- *2023-05-31* - Updated document based on the proof of concept implementation.

## Alternatives

### Runtime Assembly - Multiple pods

Test binaries are built into each release payload component image at conventionally
established location; e.g. `/openshift-tests/<component-name>`. Once a cluster
has been installed, the `openshfit-tests` binary inspects the components of the
running cluster's payload and creates pods, running those images, with
`/openshift-tests/<component-name>` binaries as the entrypoint.
`openshift-tests` would still be responsible for aggregating the results.

Pros:
- It’s “easy” to have a PR add a new test and that test will run in pre-merge testing.
- It’s easier to implement, since the coordination mechanism will exist only inside origin repository.

Cons:
- An amd64 build farm cluster cannot run release payload component pods from an ARM cluster.

> ### Build-time Assembly - Aggregating Source

Reviewer 3: there's another build-time assembly alternative:  aggregating distinct binaries into a single test image

it's the option i prefer and i'd like to see more details around why we wouldn't/can't do it that way.

Author: Added.

   main process.
7. Present the summary results of running tests.

Below diagram shows the proposed flow:

![New Flow](./new-flow.png)
<!--
Above diagram was generated at https://mermaid.live/ using following code:
sequenceDiagram
    actor U as User
    participant O as openshift-tests
    participant E as external-tests
    participant R as release:tag
    U->>+O: run openshift/conformance
    O->>+O: list tests
    O->>+R: extract test binary
    rect rgb(177, 177, 177)
        opt test binary exists?
            R->>+E: external-tests
            O->>+E: list tests
            E->>+O: tests
        end
    end
    loop over all tests
        alt embedded test?
            O->>+O: run #lt;test name#gt;#10;
        else external test?
            rect rgb(177, 177, 177)
                O->>+E: run #lt;test name#gt;
                E->>+O: results
            end
        end
    end
    O->>+U: results
-->

### API Extensions

N/A

### Implementation Details/Notes/Constraints

To achieve the first step of the proposal, we need to:
1. Provide a library for building `<binary>-tests` which exposes two commands:
   - `list` - responsible for listing tests in JSON format;

>    - `run` - run a single test, returning results in ginkgo compatible format;

Reviewer 7: Looks like I forgot the comment in a “pending” state 🙃

>    - `run` - run a single test, returning results in ginkgo compatible format;

Author: That's a good catch, thank you. The actual implementation is using `run-test`, see https://github.com/openshift/kubernetes/blob/0e0d15b865ffc36177dc8770b4723dc14476a630/openshift-hack/cmd/k8s-tests/k8s-tests.go#L64 Fix in https://github.com/openshift/enhancements/pull/1547

The proposed extensions to `openshift-tests` will require the following changes:
1. Add extraction logic responsible for retrieving test binaries from the
   release payload, based on the annotations provided by the authors of these
   images.
2. Add logic responsible for iterating over all test binaries invoking the `list`
   subcommand, and reading the list of tests provided by the binary. Every tests
   should have an associated binary name, which will be used during invocation.
3. Expand the current filtering, shuffling, and randomization logic to also take
   into account the new tests coming from external binaries.
4. Add logic responsible for iterating over all tests, invoking the `run`
   subcommand using the associated test's binary and reading its results.
5. Provide a fallback path allowing to use the vendored tests on demand (`OPENSHIFT_SKIP_EXTERNAL_TESTS`
   environment variable, for example).

### Risks and Mitigations

- **The current policy when the [origin](https://github.com/openshift/origin/) pull
  request bringing the updated k8s version breaks the CI is to revert PR. This
  policy will be harder to execute with this change because reverting the o/k

>   rebase PR can be problematic and was never tried before.**

Reviewer 8: What if the kubernetes e2e image rebase will make a current openshift/kubernetes fail? Should we temporary skip the failing tests, do the next k8s rebase and rebase the k8s e2e image?

Author: The rebase won't merge until we fix the tests, the goal is to get that signal right away when bringing in newer k8s. The only risk post-merge is that new tests rely on newer kubelet, but that case is already covered in the risks section.

  There should be fewer reasons to do any reverting, because with this new mechanism
  we should get the appropriate signal during extensive payload testing required
  to land every k8s bump PR.

- **Some of the work that has been done to build the partner certification tool will
  need to be updated to pull additional images with separate test binaries.**

  The `openshift-tests` binary will contain the necessary logic to ensure that test
  invocation is working as before.

- **Not available release image can prevent from running the tests.**

  The proposed mechanism will contain a fallback path, which will ensure that in
  cases where the release payload is not reachable we will be able to run all the
  tests bundled within the `openshift-tests` binary. This includes the basic

>   OpenShift and Kubernetes conformance tests.

Reviewer 8: Meaning the openshift/origin test suite will always have the latest k8s e2e's present?

Author: For now - yes, see [Fabio's comment](https://github.com/openshift/enhancements/pull/1291#discussion_r1246863838), and the other case is Microshift. There's a long-term goal to eventually get rid of that, but we're not there yet.

   `hyperkube` image from the release payload, but should rather rely on [an annotation](#implementation-detailsnotesconstraints),
   telling which images contain the external test binary, and its location inside
   the image.
3. Preparing a template/demo for one of operators building on top of steps 1 and 2
   above.

Additionally, to enable flows where release payload is not available, one should be
able to point to a local directory containing external binaries that should be
used instead. For example, through `OPENSHIFT_EXTERNAL_TESTS_DIRECTORY` environment
variable.

### Even more future work

As mentioned in [risks section](#risks-and-mitigations), upgrade tests are relying
on the vendored code only, it would be reasonable to address this issue in the long,
to run the upgrade tests in a similar fashion as the other tests.

Re-evaluate [goal #4](#goals) which explicitly requires providing a fallback for
kubernetes vendored tests. Currently, at least Microshift requires us to have
the kubernetes tests vendored since they don't provide release image from which
we can extract the kubernetes test binaries.

> Re-evaluate [alternative approach](#build-time-assembly---aggregating-binaries)

Reviewer 9: It's probably worth starting this conversation with testplatform in 4.15 to understand what capabilities our CI build system provides and how difficult this actually is long term.  I'd really like to see us not have vendored tests.  That will discourage separation beyond a few well known cases.

## Alternatives

### Runtime Assembly - Multiple pods

Test binaries are built into each release payload component image at conventionally
established location; e.g. `/openshift-tests/<component-name>`. Once a cluster
has been installed, the `openshfit-tests` binary inspects the components of the
running cluster's payload and creates pods, running those images, with
`/openshift-tests/<component-name>` binaries as the entrypoint.
`openshift-tests` would still be responsible for aggregating the results.

Pros:
- It’s “easy” to have a PR add a new test and that test will run in pre-merge testing.
- It’s easier to implement, since the coordination mechanism will exist only inside origin repository.

Cons:
- An amd64 build farm cluster cannot run release payload component pods from an ARM cluster.

### Build-time Assembly - Aggregating Source

`openshift-tests` contains a traditional binary for running the tests, however
it is built from a source tree composed from multiple repositories.

Cons:
- Longer test times due to the compilation of openshift-tests.
- Consistency across all repos source which contributes to this binaries.

### Build-time Assembly - Aggregating Binaries

> `openshift-tests` contains a traditional binary for running the tests, and all

Author: Done.

the additional test binaries, which are injected while building the release image.

Cons:

> - Longer test times due to the compilation of openshift-tests.

Reviewer 3: ```suggestion
- Longer test times due to the construction/aggregation of binaries from different locations into the `tests` image.
```

i assume this is what you mean.  compilation was confusing since there's no additional code compiling happening.

That said, i'd be surprised if this adds significant test time, especially since the alternative still requires:
1) building multiple individual test images (which may or may not be required with this aggregated binaries approach)
2) pulling each of those individual images at test runtime instead of pulling a single aggregate image (and then either way you need to extract each binary, either from individual images or the single image)

Author: That was copy&pasted from the above alternative - removed.

Author: I don't think that the extended time will be greater than the one we currently have, especially that we won't have to pull the binaries, since they will be there already.

> - Consistency across all repos source which contributes to this binaries.

Author: This is a limitation that we called out for all of the alternatives, so I'm not quite sure what kind of answer you're expecting. In general every split like this one will have some limitation and this particular one applies to all, afaict.

> - Consistency across all repos source which contributes to this binaries.

Author: After discussing with Ben on slack, I've removed this cons from all alternatives. Since it's rather implied limitation given the direction we're taking with this separation.

- Necessary changes in the build pipeline which will allow constructing binaries

>   from multiple repositories.

Reviewer 3: this is the challenge for sure, though i'd phrase it slightly differently.

In either scenario (runtime aggregation or buildtime) we need to build the test binaries from every repo.

In both cases we *may* also need to build a test image from each repo (build time aggregation could be done in a few ways....  one option is build the images, then extract binaries from the images and recombine them into a single image.  Another is to have some sort of build step that builds all the binaries and puts them in a single image directly.  I don't know which is easier to implement w/ our existing tools or which requires less new capability, but i'd guess the first one is probably the more obvious approach, though also probably somewhat more costly/slower to do at build time since you're building/pushing/pulling extra images)

The clear new capability/change needed in the pipeline is the ability to include content from either multiple images, or multiple repos, in a single payload image.

Note also that we could do a super hacky thing which requires no pipeline changes, if we hardcode the image layers.  e.g. the `tests` Dockerfile can just contain a bunch of "FROM" stages that pulls the test binaries from every team's individual test image.  Of course this requires teams go update the `tests` Dockerfile whenever they want to introduce a new component level tests image.  But it really might not be that bad...it's not like it's going to change that often.  We would just need to ensure that when doing PR presubmit jobs, for example, the pipeline correctly recognizes it needs to rebuild the aggregate `tests` image when any of the component images are touched by the PR.  I assume it already knows how to do that, though.

Author: I've updated the wording, but I wouldn't want to focus on solving alternative solutions, yet. I've added that as a future work, and I'd prefer to leave the decision towards the approach until then. Especially that between now and then our pipeline's functionality might and will change, maybe providing even simpler way to solve that particular problem.
